// eventregistry/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var analytics_1 = require("./analytics");
exports.Analytics = analytics_1.Analytics;
var base_1 = require("./base");
exports.QueryItems = base_1.QueryItems;
var counts_1 = require("./counts");
exports.GetCounts = counts_1.GetCounts;
exports.GetCountsEx = counts_1.GetCountsEx;
var dailyShares_1 = require("./dailyShares");
exports.GetTopSharedArticles = dailyShares_1.GetTopSharedArticles;
exports.GetTopSharedEvents = dailyShares_1.GetTopSharedEvents;
var eventForText_1 = require("./eventForText");
exports.GetEventForText = eventForText_1.GetEventForText;
var eventRegistry_1 = require("./eventRegistry");
exports.EventRegistry = eventRegistry_1.EventRegistry;
exports.ArticleMapper = eventRegistry_1.ArticleMapper;
var base_2 = require("./base");
exports.sleep = base_2.sleep;
exports.mainLangs = base_2.mainLangs;
exports.allLangs = base_2.allLangs;
var info_1 = require("./info");
exports.GetSourceInfo = info_1.GetSourceInfo;
exports.GetConceptInfo = info_1.GetConceptInfo;
exports.GetCategoryInfo = info_1.GetCategoryInfo;
exports.GetSourceStats = info_1.GetSourceStats;
var query_1 = require("./query");
exports.BaseQuery = query_1.BaseQuery;
exports.CombinedQuery = query_1.CombinedQuery;
exports.ComplexArticleQuery = query_1.ComplexArticleQuery;
exports.ComplexEventQuery = query_1.ComplexEventQuery;
var queryArticle_1 = require("./queryArticle");
exports.QueryArticle = queryArticle_1.QueryArticle;
exports.RequestArticleInfo = queryArticle_1.RequestArticleInfo;
exports.RequestArticleSimilarArticles = queryArticle_1.RequestArticleSimilarArticles;
exports.RequestArticleDuplicatedArticles = queryArticle_1.RequestArticleDuplicatedArticles;
exports.RequestArticleOriginalArticle = queryArticle_1.RequestArticleOriginalArticle;
var queryArticles_1 = require("./queryArticles");
exports.QueryArticles = queryArticles_1.QueryArticles;
exports.QueryArticlesIter = queryArticles_1.QueryArticlesIter;
exports.RequestArticlesInfo = queryArticles_1.RequestArticlesInfo;
exports.RequestArticlesUriWgtList = queryArticles_1.RequestArticlesUriWgtList;
exports.RequestArticlesTimeAggr = queryArticles_1.RequestArticlesTimeAggr;
exports.RequestArticlesConceptAggr = queryArticles_1.RequestArticlesConceptAggr;
exports.RequestArticlesCategoryAggr = queryArticles_1.RequestArticlesCategoryAggr;
exports.RequestArticlesSourceAggr = queryArticles_1.RequestArticlesSourceAggr;
exports.RequestArticlesKeywordAggr = queryArticles_1.RequestArticlesKeywordAggr;
exports.RequestArticlesConceptGraph = queryArticles_1.RequestArticlesConceptGraph;
exports.RequestArticlesConceptMatrix = queryArticles_1.RequestArticlesConceptMatrix;
exports.RequestArticlesConceptTrends = queryArticles_1.RequestArticlesConceptTrends;
exports.RequestArticlesDateMentionAggr = queryArticles_1.RequestArticlesDateMentionAggr;
exports.RequestArticlesRecentActivity = queryArticles_1.RequestArticlesRecentActivity;
var queryEvent_1 = require("./queryEvent");
exports.QueryEvent = queryEvent_1.QueryEvent;
exports.QueryEventArticlesIter = queryEvent_1.QueryEventArticlesIter;
exports.RequestEventInfo = queryEvent_1.RequestEventInfo;
exports.RequestEventArticles = queryEvent_1.RequestEventArticles;
exports.RequestEventArticleUriWgts = queryEvent_1.RequestEventArticleUriWgts;
exports.RequestEventKeywordAggr = queryEvent_1.RequestEventKeywordAggr;
exports.RequestEventSourceAggr = queryEvent_1.RequestEventSourceAggr;
exports.RequestEventDateMentionAggr = queryEvent_1.RequestEventDateMentionAggr;
exports.RequestEventArticleTrend = queryEvent_1.RequestEventArticleTrend;
exports.RequestEventSimilarEvents = queryEvent_1.RequestEventSimilarEvents;
var queryEvents_1 = require("./queryEvents");
exports.QueryEvents = queryEvents_1.QueryEvents;
exports.QueryEventsIter = queryEvents_1.QueryEventsIter;
exports.RequestEventsInfo = queryEvents_1.RequestEventsInfo;
exports.RequestEventsUriWgtList = queryEvents_1.RequestEventsUriWgtList;
exports.RequestEventsTimeAggr = queryEvents_1.RequestEventsTimeAggr;
exports.RequestEventsKeywordAggr = queryEvents_1.RequestEventsKeywordAggr;
exports.RequestEventsLocAggr = queryEvents_1.RequestEventsLocAggr;
exports.RequestEventsLocTimeAggr = queryEvents_1.RequestEventsLocTimeAggr;
exports.RequestEventsConceptAggr = queryEvents_1.RequestEventsConceptAggr;
exports.RequestEventsConceptGraph = queryEvents_1.RequestEventsConceptGraph;
exports.RequestEventsConceptMatrix = queryEvents_1.RequestEventsConceptMatrix;
exports.RequestEventsConceptTrends = queryEvents_1.RequestEventsConceptTrends;
exports.RequestEventsSourceAggr = queryEvents_1.RequestEventsSourceAggr;
exports.RequestEventsDateMentionAggr = queryEvents_1.RequestEventsDateMentionAggr;
exports.RequestEventsEventClusters = queryEvents_1.RequestEventsEventClusters;
exports.RequestEventsCategoryAggr = queryEvents_1.RequestEventsCategoryAggr;
exports.RequestEventsRecentActivity = queryEvents_1.RequestEventsRecentActivity;
var queryStory_1 = require("./queryStory");
exports.QueryStory = queryStory_1.QueryStory;
exports.RequestStoryInfo = queryStory_1.RequestStoryInfo;
exports.RequestStoryArticles = queryStory_1.RequestStoryArticles;
exports.RequestStoryArticleUris = queryStory_1.RequestStoryArticleUris;
exports.RequestStoryArticleTrend = queryStory_1.RequestStoryArticleTrend;
exports.RequestStorySimilarStories = queryStory_1.RequestStorySimilarStories;
var recent_1 = require("./recent");
exports.GetRecentEvents = recent_1.GetRecentEvents;
exports.GetRecentArticles = recent_1.GetRecentArticles;
var returnInfo_1 = require("./returnInfo");
exports.ReturnInfo = returnInfo_1.ReturnInfo;
exports.ArticleInfoFlags = returnInfo_1.ArticleInfoFlags;
exports.StoryInfoFlags = returnInfo_1.StoryInfoFlags;
exports.EventInfoFlags = returnInfo_1.EventInfoFlags;
exports.SourceInfoFlags = returnInfo_1.SourceInfoFlags;
exports.CategoryInfoFlags = returnInfo_1.CategoryInfoFlags;
exports.ConceptInfoFlags = returnInfo_1.ConceptInfoFlags;
exports.LocationInfoFlags = returnInfo_1.LocationInfoFlags;
exports.ConceptClassInfoFlags = returnInfo_1.ConceptClassInfoFlags;
exports.ConceptFolderInfoFlags = returnInfo_1.ConceptFolderInfoFlags;
var trends_1 = require("./trends");
exports.GetTrendingConcepts = trends_1.GetTrendingConcepts;
exports.GetTrendingCategories = trends_1.GetTrendingCategories;
exports.GetTrendingCustomItems = trends_1.GetTrendingCustomItems;
exports.GetTrendingConceptGroups = trends_1.GetTrendingConceptGroups;
var topicPage_1 = require("./topicPage");
exports.TopicPage = topicPage_1.TopicPage;
