// eventregistry/dist/topicPage.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var _ = require("lodash");
var base_1 = require("./base");
var returnInfo_1 = require("./returnInfo");
var TopicPage = /** @class */ (function (_super) {
    __extends(TopicPage, _super);
    function TopicPage(eventRegistry) {
        var _this = _super.call(this) || this;
        _this.eventRegistry = eventRegistry;
        _this.emptyTopicPage = {
            autoAddArticles: true,
            articleHasDuplicate: "keepAll",
            articleHasEvent: "keepAll",
            articleIsDuplicate: "skipDuplicates",
            maxDaysBack: 7,
            articleTreshWgt: 0,
            eventTreshWgt: 0,
            concepts: [],
            keywords: [],
            categories: [],
            sources: [],
            sourceGroups: [],
            sourceLocations: [],
            locations: [],
            langs: [],
            restrictToSetConcepts: false,
            restrictToSetCategories: false,
            restrictToSetSources: false,
            restrictToSetLocations: false,
            dataType: ["news"]
        };
        _this.concept = {};
        _this.topicPage = _this.createEmptyTopicPage();
        return _this;
    }
    TopicPage.prototype.isTopicPage = function (definition) {
        return _.isEqual(_.keys(this.emptyTopicPage), _.keys(definition));
    };
    TopicPage.prototype.createEmptyTopicPage = function () {
        return this.emptyTopicPage;
    };
    /**
     * Load an existing topic page from Event Registry based on the topic page URI
     * @param uri uri of the topic page saved in your Event Registry account
     */
    TopicPage.prototype.loadTopicPageFromER = function (uri) {
        return __awaiter(this, void 0, void 0, function () {
            var params, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        params = {
                            action: "getTopicPageJson",
                            includeConceptDescription: true,
                            includeConceptImage: true,
                            includeTopicPageDefinition: true,
                            includeTopicPageOwner: true,
                            uri: uri
                        };
                        this.topicPage = this.createEmptyTopicPage();
                        _a = this;
                        return [4 /*yield*/, this.eventRegistry.jsonRequest("/api/v1/topicPage", params)];
                    case 1:
                        _a.concept = _b.sent();
                        if (_.has(this.concept, "data.error")) {
                            throw new Error(_.get(this.concept, "data.error", ""));
                        }
                        this.topicPage = _.extend({}, this.topicPage, _.get(this.concept, "data.topicPage", {}));
                        return [2 /*return*/, this.topicPage];
                }
            });
        });
    };
    /**
     * Load the topic page definition from an object
     * @param definition topic page definition
     */
    TopicPage.prototype.loadTopicPageFromDefinition = function (definition) {
        if (this.isTopicPage(definition)) {
            this.topicPage = definition;
        }
    };
    /**
     * Load topic page from an existing file
     * @param filepath file that contains the topic page definition
     */
    TopicPage.prototype.loadTopicPageFromFile = function (filepath) {
        if (fs && fs.existsSync(filepath)) {
            var definition = JSON.parse(fs.readFileSync(filepath, "utf8"));
            if (this.isTopicPage(definition)) {
                this.topicPage = definition;
            }
        }
    };
    /**
     * Return an object containing the topic page definition. you can use it to load a topic page later
     */
    TopicPage.prototype.saveTopicPageDefinition = function () {
        return this.topicPage;
    };
    /**
     * Save the topic page definition to a file
     * @param filepath location of the file to which you'll save the topic page definition
     */
    TopicPage.prototype.saveTopicPageDefinitionToFile = function (filepath) {
        fs.writeFileSync(filepath, JSON.stringify(this.topicPage, null, 4), "utf8");
    };
    Object.defineProperty(TopicPage.prototype, "articleThreshold", {
        /**
         * What is the minimum total weight that an article has to have in order to get it among the results?
         */
        set: function (value) {
            if (_.isNumber(value) && value >= 0) {
                this.topicPage.articleTreshWgt = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "eventThreshold", {
        /**
         * What is the minimum total weight that an event has to have in order to get it among the results?
         */
        set: function (value) {
            if (_.isNumber(value) && value >= 0) {
                this.topicPage.eventTreshWgt = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "articleIsDuplicateFilter", {
        /**
         * Some articles can be duplicates of other articles. What should be done with them. Possible values are:
         * "skipDuplicates" (skip the resulting articles that are duplicates of other articles),
         * "keepOnlyDuplicates" (return only the duplicate articles) and
         * "keepAll" (no filtering, default)
         */
        set: function (value) {
            if (value === "skipDuplicates" || value === "keepOnlyDuplicates" || value === "keepAll") {
                this.topicPage.articleIsDuplicate = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "articleHasEventFilter", {
        /**
         * Some articles describe a known event and some don't. This filter allows you to filter the resulting articles based on this criteria.
         * Possible values are:
         * "skipArticlesWithoutEvent" (skip articles that are not describing any known event in ER),
         * "keepOnlyArticlesWithoutEvent" (return only the articles that are not describing any known event in ER) and
         * "keepAll" (no filtering, default)
         */
        set: function (value) {
            if (value === "skipArticlesWithoutEvent" || value === "keepOnlyArticlesWithoutEvent" || value === "keepAll") {
                this.topicPage.articleHasEvent = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "articleHasDuplicateFilter", {
        /**
         * Some articles are later copied by others. What should be done with such articles. Possible values are:
         * "skipHasDuplicates" (skip the articles that have been later copied by others),
         * "keepOnlyHasDuplicates" (return only the articles that have been later copied by others) and
         * "keepAll" (no filtering, default)
         */
        set: function (value) {
            if (value === "skipHasDuplicates" || value === "keepOnlyHasDuplicates" || value === "keepAll") {
                this.topicPage.articleHasDuplicate = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "dataTypes", {
        /**
         * What data types should we search? "news" (news content, default), "pr" (press releases), or "blog".
         * If you want to use multiple data types, put them in an array (e.g. ["news", "pr"])
         */
        set: function (dataTypes) {
            this.topicPage.dataType = dataTypes;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "sourceRankStartPercentile", {
        set: function (startPercentile) {
            if (_.isUndefined(startPercentile)) {
                startPercentile = 0;
            }
            if (startPercentile < 0 || startPercentile > 90) {
                throw new RangeError("startPercentile is out of valid values (0 - 90)");
            }
            if (!_.isNil(this.topicPage.endSourceRankPercentile) && _.isNumber(this.topicPage.endSourceRankPercentile) && startPercentile >= this.topicPage.endSourceRankPercentile) {
                throw new RangeError("startPercentile has to be smaller than endPercentile");
            }
            if (startPercentile % 10 !== 0) {
                throw new Error("startPecentile has to be a multiple of 10");
            }
            this.topicPage.startSourceRankPercentile = startPercentile;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "sourceRankEndPercentile", {
        set: function (endPercentile) {
            if (_.isUndefined(endPercentile)) {
                endPercentile = 100;
            }
            if (endPercentile < 10 || endPercentile > 100) {
                throw new RangeError("endPercentile is out of valid values (10 - 100)");
            }
            if (!_.isNil(this.topicPage.startSourceRankPercentile) && _.isNumber(this.topicPage.startSourceRankPercentile) && this.topicPage.startSourceRankPercentile >= endPercentile) {
                throw new RangeError("startPercentile has to be smaller than endPercentile");
            }
            if (endPercentile % 10 !== 0) {
                throw new Error("endPercentile has to be a multiple of 10");
            }
            this.topicPage.endSourceRankPercentile = endPercentile;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "maxDaysBack", {
        /**
         * What is the maximum allowed age of the results?
         */
        set: function (maxDaysBack) {
            if (!_.isNumber(maxDaysBack)) {
                throw new Error("maxDaysBack value has to be a positive number");
            }
            if (maxDaysBack <= 0) {
                throw new RangeError("maxDaysBack value has to be a positive number");
            }
            this.topicPage.maxDaysBack = maxDaysBack;
        },
        enumerable: true,
        configurable: true
    });
    TopicPage.prototype.clearConcepts = function () {
        this.topicPage.concepts = [];
    };
    TopicPage.prototype.clearKeywords = function () {
        this.topicPage.keywords = [];
    };
    TopicPage.prototype.clearCategories = function () {
        this.topicPage.categories = [];
    };
    TopicPage.prototype.clearSources = function () {
        this.topicPage.sources = [];
    };
    TopicPage.prototype.clearSourceLocations = function () {
        this.topicPage.sourceLocations = [];
    };
    TopicPage.prototype.clearSourceGroups = function () {
        this.topicPage.sourceGroups = [];
    };
    TopicPage.prototype.clearLocations = function () {
        this.topicPage.locations = [];
    };
    /**
     * Add a relevant concept to the topic page
     * @param uri uri of the concept to be added
     * @param weight importance of the provided concept (typically in range 1 - 50)
     */
    TopicPage.prototype.addConcept = function (uri, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var label = args.label, conceptType = args.conceptType, _a = args.required, required = _a === void 0 ? false : _a, _b = args.excluded, excluded = _b === void 0 ? false : _b;
        if (required && excluded) {
            throw new Error("Parameters required and excluded can not be true at the same time");
        }
        var concept = { uri: uri, wgt: wgt, required: required, excluded: excluded };
        if (!_.isUndefined(label)) {
            _.set(concept, "label", label);
        }
        if (!_.isUndefined(conceptType)) {
            _.set(concept, "type", conceptType);
        }
        this.topicPage.concepts = __spread(this.topicPage.concepts, [concept]);
    };
    /**
     * Add a relevant keyword to the topic page
     * @param keyword keyword or phrase to be added
     * @param weight importance of the provided keyword (typically in range 1 - 50)
     */
    TopicPage.prototype.addKeyword = function (keyword, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var _a = args.required, required = _a === void 0 ? false : _a, _b = args.excluded, excluded = _b === void 0 ? false : _b;
        this.topicPage.keywords = __spread(this.topicPage.keywords, [{ keyword: keyword, wgt: wgt, required: required, excluded: excluded }]);
    };
    /**
     * Add a relevant category to the topic page
     * @param uri uri of the category to be added
     * @param weight importance of the provided category (typically in range 1 - 50)
     */
    TopicPage.prototype.addCategory = function (uri, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var _a = args.required, required = _a === void 0 ? false : _a, _b = args.excluded, excluded = _b === void 0 ? false : _b;
        this.topicPage.categories = __spread(this.topicPage.categories, [{ uri: uri, wgt: wgt, required: required, excluded: excluded }]);
    };
    /**
     * Add a news source to the topic page
     * @param uri uri of the news source to add to the topic page
     * @param weight importance of the news source (typically in range 1 - 50)
     */
    TopicPage.prototype.addSource = function (uri, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var _a = args.excluded, excluded = _a === void 0 ? false : _a;
        this.topicPage.sources = __spread(this.topicPage.sources, [{ uri: uri, wgt: wgt, excluded: excluded }]);
    };
    /**
     * Add a list of relevant sources by identifying them by their geographic location
     * @param uri uri of the location where the sources should be geographically located
     * @param weight importance of the provided list of sources (typically in range 1 - 50)
     */
    TopicPage.prototype.addSourceLocation = function (uri, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var _a = args.excluded, excluded = _a === void 0 ? false : _a;
        this.topicPage.sourceLocations = __spread(this.topicPage.sourceLocations, [{ uri: uri, wgt: wgt, excluded: excluded }]);
    };
    /**
     * Add a list of relevant sources by specifying a whole source group to the topic page
     * @param sourceGroupUri uri of the source group to add
     * @param weight importance of the provided list of sources (typically in range 1 - 50)
     */
    TopicPage.prototype.addSourceGroup = function (uri, wgt, args) {
        if (args === void 0) { args = {}; }
        if (!_.isNumber(wgt)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        var _a = args.excluded, excluded = _a === void 0 ? false : _a;
        this.topicPage.sourceGroups = __spread(this.topicPage.sourceGroups, [{ uri: uri, wgt: wgt, excluded: excluded }]);
    };
    /**
     * Add relevant location to the topic page
     * @param locationUri uri of the location to add
     * @param weight importance of the provided location (typically in range 1 - 50)
     */
    TopicPage.prototype.addLocation = function (locationUri, weight) {
        if (!_.isNumber(weight)) {
            throw new Error("Weight value has to be a positive or negative number");
        }
        this.topicPage.locations = __spread(this.topicPage.locations, [{ uri: locationUri, wgt: weight }]);
    };
    Object.defineProperty(TopicPage.prototype, "languages", {
        /**
         * Restrict the results to the list of specified languages
         */
        set: function (languages) {
            if (!_.isArray(languages)) {
                languages = [languages];
            }
            if (_.every(languages, function (language) { return _.size(language) === 3; })) {
                this.topicPage.langs = languages;
            }
            else {
                throw new Error("Expected to get language in ISO3 code");
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "restrictToSetConceptsAndKeywords", {
        /**
         * if true then the results have to mention at least one of the specified concepts or keywords
         */
        set: function (restrict) {
            if (!_.isBoolean(restrict)) {
                throw new Error("Restrict value has to be a boolean value");
            }
            this.topicPage.restrictToSetConcepts = restrict;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "restrictToSetCategories", {
        /**
         * if set to true then return only results that are assigned to one of the specified categories
         */
        set: function (restrict) {
            if (!_.isBoolean(restrict)) {
                throw new Error("Restrict value has to be a boolean value");
            }
            this.topicPage.restrictToSetCategories = restrict;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "restrictToSetSources", {
        /**
         * if set to true then return only results from one of the specified news sources
         * this includes also sources set by source groups or by source locations
         */
        set: function (restrict) {
            if (!_.isBoolean(restrict)) {
                throw new Error("Restrict value has to be a boolean value");
            }
            this.topicPage.restrictToSetSources = restrict;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TopicPage.prototype, "restrictToSetLocations", {
        /**
         * if set to true, then return only results that are located at one of the specified locations
         */
        set: function (restrict) {
            if (!_.isBoolean(restrict)) {
                throw new Error("Restrict value has to be a boolean value");
            }
            this.topicPage.restrictToSetLocations = restrict;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Return a list of articles that match the topic page
     * @param args {EventRegistryStatic.TopicPageArticles} Object which contains a host of optional parameters
     */
    TopicPage.prototype.getArticles = function (args) {
        if (args === void 0) { args = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, page, _b, count, _c, sortBy, _d, sortByAsc, _e, dataType, _f, returnInfo, otherParameters, params, request;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        _a = args.page, page = _a === void 0 ? 1 : _a, _b = args.count, count = _b === void 0 ? 100 : _b, _c = args.sortBy, sortBy = _c === void 0 ? "rel" : _c, _d = args.sortByAsc, sortByAsc = _d === void 0 ? false : _d, _e = args.dataType, dataType = _e === void 0 ? "news" : _e, _f = args.returnInfo, returnInfo = _f === void 0 ? new returnInfo_1.ReturnInfo() : _f, otherParameters = __rest(args, ["page", "count", "sortBy", "sortByAsc", "dataType", "returnInfo"]);
                        if (page < 1) {
                            throw new RangeError("page has to be >= 1");
                        }
                        if (count > 100) {
                            throw new RangeError("At most 100 articles can be returned at a time");
                        }
                        params = _.extend({}, {
                            action: "getArticlesForTopicPage",
                            resultType: "articles",
                            dataType: this.topicPage.dataType,
                            articlesCount: count,
                            articlesSortBy: sortBy,
                            articlesSortByAsc: sortByAsc,
                            articlesPage: page,
                            topicPage: JSON.stringify(this.topicPage),
                        }, returnInfo.getParams());
                        if (!_.isEmpty(otherParameters)) {
                            params = __assign({}, params, otherParameters);
                        }
                        return [4 /*yield*/, this.eventRegistry.jsonRequest("/api/v1/article", params)];
                    case 1:
                        request = _g.sent();
                        return [2 /*return*/, _.get(request, "data.articles.results", [])];
                }
            });
        });
    };
    /**
     * Return a list of events that match the topic page
     * @param args {EventRegistryStatic.TopicPageEvents} Object which contains a host of optional parameters
     */
    TopicPage.prototype.getEvents = function (args) {
        if (args === void 0) { args = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, page, _b, count, _c, sortBy, _d, sortByAsc, _e, returnInfo, otherParameters, params, request;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _a = args.page, page = _a === void 0 ? 1 : _a, _b = args.count, count = _b === void 0 ? 50 : _b, _c = args.sortBy, sortBy = _c === void 0 ? "rel" : _c, _d = args.sortByAsc, sortByAsc = _d === void 0 ? false : _d, _e = args.returnInfo, returnInfo = _e === void 0 ? new returnInfo_1.ReturnInfo() : _e, otherParameters = __rest(args, ["page", "count", "sortBy", "sortByAsc", "returnInfo"]);
                        if (page < 1) {
                            throw new RangeError("page has to be >= 1");
                        }
                        if (count > 50) {
                            throw new RangeError("At most 50 events can be returned at a time");
                        }
                        params = _.extend({}, {
                            action: "getEventsForTopicPage",
                            resultType: "events",
                            dataType: this.topicPage.dataType,
                            eventsCount: count,
                            eventsSortBy: sortBy,
                            eventsSortByAsc: sortByAsc,
                            eventsPage: page,
                            topicPage: JSON.stringify(this.topicPage),
                        }, returnInfo.getParams());
                        if (!_.isEmpty(otherParameters)) {
                            params = __assign({}, params, otherParameters);
                        }
                        return [4 /*yield*/, this.eventRegistry.jsonRequest("/api/v1/event", params)];
                    case 1:
                        request = _f.sent();
                        return [2 /*return*/, _.get(request, "data.events.results", [])];
                }
            });
        });
    };
    return TopicPage;
}(base_1.QueryParamsBase));
exports.TopicPage = TopicPage;
